[WebAssembly] [Backend] Wasm optimize illegal bitmask

badumbatish opened the pull request:
[WebAssembly] [Backend] Wasm optimize illegal bitmask for #131980.

Currently, the case for illegal bitmask (v32i8 or v64i8) is that at the SelectionDag level,  two (four) vectors of v128 will be concatenated together, then they'll all be SETCC by the same pseudo illegal instruction, which requires expansion later on.

I opt for SETCC-ing them seperately, bitcast and zext them and then add them up together in the end.

badumbatish committed:
[WebAssembly] Precommit test for optimizing illegal bitmask

badumbatish committed:
[WebAssembly] [Backend] Optimize illegal bitmask

badumbatish commented:
@lukel97  hi Luke! just tagging you in just in case you find it interesting

badumbatish commented:
i'm finding out that in web assembly, in llvm ir, v64i8 can be converted to 4 v128 as input to the selection dag, but v64i4 will degenerate to 64 integers of i32. I'm not sure how to handle this case so that v64i4 can be converted to 2 v128? 


I set it via WebAssembly::getPreferredVectorAction so that if (EltVT == MVT::v64i4) return TypeSplitVector;

the only place that calls this function in TargetLoweringBase is computeRegisterProperties.

The enum fall throughs to TypeScalarizeVector (below is the block for it in computeRegisterProperties), but after this i'm not sure where to go next

    case TypeScalarizeVector: {
      MVT IntermediateVT;
      MVT RegisterVT;
      unsigned NumIntermediates;
      unsigned NumRegisters = getVectorTypeBreakdownMVT(VT, IntermediateVT,
          NumIntermediates, RegisterVT, this);
      NumRegistersForVT[i] = NumRegisters;
      assert(NumRegistersForVT[i] == NumRegisters &&
             "NumRegistersForVT size cannot represent NumRegisters!");
      RegisterTypeForVT[i] = RegisterVT;

      MVT NVT = VT.getPow2VectorType();
      if (NVT == VT) {
        // Type is already a power of 2.  The default action is to split.
        TransformToType[i] = MVT::Other;
        if (PreferredAction == TypeScalarizeVector)
          ValueTypeActions.setTypeAction(VT, TypeScalarizeVector);
        else if (PreferredAction == TypeSplitVector)
          ValueTypeActions.setTypeAction(VT, TypeSplitVector);
        else if (EC.getKnownMinValue() > 1)
          ValueTypeActions.setTypeAction(VT, TypeSplitVector);
        else
          ValueTypeActions.setTypeAction(VT, EC.isScalable()
                                                 ? TypeScalarizeScalableVector
                                                 : TypeScalarizeVector);
      } else {
        TransformToType[i] = NVT;
        ValueTypeActions.setTypeAction(VT, TypeWidenVector);
      }
      break;
    }

badumbatish commented:
@sparker-arm hi there! i'm not sure who to tag for this issue I'm having, but would it be possible for a <64 x i4> to be converted to 2 v128 since <64 x i8> can be converted to 4 v128? Right now without converting to v128, the assembly degenerates to about 400 lines. 

If its possible to convert <64 x i4> to 2 v128, I'd appreciate some pointers or some directions on where to proceed next

sparker-arm commented:
Out of curiosity, are there any languages, or frameworks, compiling to wasm that will use an i4? Either way, I think it would be okay, and easier to review, if i4 support came in a separate patch.

badumbatish commented:
> Out of curiosity, are there any languages, or frameworks, compiling to wasm that will use an i4? Either way, I think it would be okay, and easier to review, if i4 support came in a separate patch.

hmm not that i know of, the llvm ir snippet that has <64 x i4 > is hand written by me, I can remove it if it's not applicable?

sparker-arm commented:
> I can remove it if it's not applicable?

Keep the test in, I imagine we'll have to deal with int4 at some point! Code LGTM anyway, thanks. Let's get a sign off from someone else though.

Commit:
[WebAssembly] Add varying vector precommittest

Add nonconst, varying vector precommit test to simd-illegal-bitmask test.

Commit:
[WebAssembly] Add support for illgal varying vect

Added bitmask optimizing for illegal varying vector. Before we can only
optimize const vector in setcc.

badumbatish commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3236,6 +3244,48 @@ static SDValue performBitcastCombine(SDNode *N,
+      return SDValue();
+    // CHECK IF VECTOR is a constant, i.e all values are the same
+    if (!ISD::isBuildVectorOfConstantSDNodes(SetCCVector.getNode()))
+      return SDValue();
```
> I'm also wondering is there a reason why SetCCVector needs to be a splat to begin with? Could we allow non-constant, non-splat vectors if we extracted a v16i8 subvector instead with ISD::extract_subvector?

I think it started as a prototype first to see if the idea works. I've added two more tests for this case.

dschuff reviewed (approved):
LGTM from the wasm side.

badumbatish commented:
thanks Derek, i don't have commit access right now. Would love for somebody to help me land the patch

dschuff commented:
I can do that. Can you update the OP comment with the text you want to have as the commit message? Currently it still says "draft".

badumbatish commented:
> I can do that. Can you update the OP comment with the text you want to have as the commit message? Currently it still says "draft".

changed. Tysm!
